### ISSUE TITLE
"Unknown error" when trying to backup files with path length >260 characters

### ISSUE THREAD

**Issue by 4001982248998** [grke/burp#807]
When backing up Windows filesystems with path names that exceed 260 characters, it will throw an “Unknown error” like this (sic):

```
2019-04-01 16:02:03 +0200: burp[24483] WARNING: Could not stat C:/test/This is an example directory that has an extreemly long name that is more than 248 characters in length to serve as an example of how to go beyond the normal limit - Note that you will not be able to see it in Windows Explorer due to the f/This is an example directory that has an extreemly long name that is more than 248 characters in length to serve as an example of how to go beyond the normal limit - Note that you will not be able to see it in Windows Explorer due to the:
```
or, more realistically:
```
2019-04-01 15:36:12 +0200: burp[19985] WARNING: Could not stat d:/xxxxxxxxxxxxx/Profile/Surname.Lastname.V2/AppData/Roaming/Macromedia/Flash Player/#SharedObjects/94CSXQTX/www.tagesschau.de/[[IMPORT]]/players.edgesuite.net/flash/plugins/osmf/advanced-streaming-plugin/v3.1/osmf2.0/AkamaiAdvancedStreamingPlugin.swf/HDCore.sol: Unknown error
```

This seems to correspond to the 260-character-limitation imposed by ``MAX_PATH`` described [here](https://docs.microsoft.com/de-de/windows/desktop/FileIO/naming-a-file#paths).

The situation can be reproduced like that:

  1. create a directory with a name like ``a``
  0. in this directory, create a subdirectory with a long name like ``This is an example directory that has an extreemly long name that is more than 248 characters in length to serve as an example of how to go beyond the normal limit - Note that you will not be able to see it in Windows Explorer due to the fact that it is limited to displaying files with fewer than 260 characters in the name``
  0. rename the parent directory from ``a`` to ``This is an example directory that has an extreemly long name that is more than 248 characters in length to serve as an example of how to go beyond the normal limit - Note that you will not be able to see it in Windows Explorer due to the fact that it is limited to displaying files with fewer than 260 characters in the name``, so now there a two nested directories with ridiculously long path names.

I see two aspects in which the burp's behaviour could be improved:

1. The algorithm should recognize the cause of the error and change the description from ``Unknown error`` to ``max_path limitation exceeded`` or something similar
2. As clients are somehow handle data on the server, at least creating these files, they probably should be backed up, and not have to be excluded from the backup “due to technical reasons”. I currently have handle this with the option ``exclude_regex = ^.{260,}$`` and thereby excluding everything with path lengths greater than 260 chars, but this does not seem optimal to me.

**Comment by grke** [comment]
Hello,
Thank you for taking the time to write up your observations, I will take a look at this when I have time.

**Comment by deajan** [comment]
This is a known Windows limitation, that can only be overcome by using UNC paths which allow 32767 characters instead of 260. I used to have a lot of backup software in dark ages that had this flaw.

Indeed, most system tools / api support UNC paths, eg the following ```dir \\?\C:\Users``` or ```echo test >> \\?\C:\somefile.txt``` work well on Win7 upwards (maybe even on Windows XP SP3 ?).

@grke: Is it possible to prefix all system file calls with "\\?\" but remove that prefix from the actual storage and manifest, so regexes and file lists stay usable ? Restores could then also benefit from a "\\?\" suffix.

**Comment by grke** [comment]
Most of the burp src/win32 stuff does make an effort to do the \\?\ thing, as far as I understand (this is code inherited from bacula).
However, some change or another that I have done has probably broken it.

**Comment by grke** [comment]
Hello,

I do not know how to create the long directory names for your test.

I tried on Windows 7 and Windows 10, using Windows Explorer, and also from a cmd prompt, and from Powershell. I've tried adding a "LongPathsEnabled" key to the registry and rebooting, and that also doesn't let me create long paths.

Are you able to explain to me how to create these directories on either Windows 7 or Windows 10?

**Comment by 4001982248998** [comment]
Hej,

I _think_ you can do it like I described it above:

>     1. create a directory with a name like `a`
>     2. in this directory, create a subdirectory with a long name like `This is an example directory that has an extreemly long name that is more than 248 characters in length to serve as an example of how to go beyond the normal limit - Note that you will not be able to see it in Windows Explorer due to the fact that it is limited to displaying files with fewer than 260 characters in the name`
>     3. rename the parent directory from `a` to `This is an example directory that has an extreemly long name that is more than 248 characters in length to serve as an example of how to go beyond the normal limit - Note that you will not be able to see it in Windows Explorer due to the fact that it is limited to displaying files with fewer than 260 characters in the name`, so now there a two nested directories with ridiculously long path names.

I tried to demo this in a video which I uploaded to https://streamable.com/3r6ul.

**Comment by deajan** [comment]
You may use this script that will create a 281 chars long path in c:\ which recursively contains 8 directories of 32 characters each, the last directory containing a file.

Just make sure temp_path is not an existing drive (replace R: with whatever if R: already exists).


```
::@echo off

SETLOCAL EnableDelayedExpansion

:: We need to create nested directories, ie 8x32 (256) + filename

set create_path=C:
set temp_path=R:
set chars32=abcdefghijklmnopqrstuvwxyz123456

%create_path%
cd\

:: Create 32 x 7 char long path

for /L %%i IN (1,1,7) do (
mkdir %chars32%
cd %chars32%
)

:: Now fool OS limit

subst %temp_path% .
%temp_path%
mkdir %chars32%
cd %chars32%

echo Hello, UNC world> "some file.txt"

C:
subst %temp_path% /D
```

**Comment by grke** [comment]
> I tried to demo this in a video which I uploaded to https://streamable.com/3r6ul.

Oh, that's easier to understand. You're not actually pasting the full path that you quoted above.
I will try again soon.

**Comment by grke** [comment]
And deajan's suggestion looks helpful too, thanks!

**Comment by grke** [comment]
I've been trying to reproduce this again, but haven't managed it yet.
I created lots of subdirectories and a file 300+ characters deep and burp backed up and restored it fine.
I am using the latest burp 2.3.10.

Now I'm wondering what burp version this behaviour is being seen on?

**Comment by deajan** [comment]
Hmm... Also tried with my example script, could backup / restore without any problem.
My test was a 281 char long path that got restored into another directory giving a 289 char long restore path.
Using burp 2.3.6 on Win10 x64, tried protocol 1.
@4001982248998 Do you have any more context ? Like Windows version ? Maybe test with a newer Windows too to see if it's WinXP/7/8/10 version related.  What burp version are you using ?

**Comment by 4001982248998** [comment]
Hi, thank you for coming back to this!

My server and client versions are 2.2.18, using protocol version 1.

Windows servers have version Windows Server 2016, Version 1607, Build 14393.3085.

Reproducing the problem is hard now — the troublesome directories have since been administratively changed so that they don't exceed the maximum character number. I will try nonetheless.

**Comment by grke** [comment]
Hello,
(Using burp 2.3.10...)
Although I haven't reproduced an issue on backing up long paths, whilst trying to reproduce it, I found an issue when restoring long paths.
I think I have fixed it now - it was really hard to figure out!

**Comment by grke** [comment]
The restore fix will be merged to master soon, if it passes all my automatic tests. I am putting some long paths into the automatic tests that I run too, to try avoid breaking it again.

**Comment by grke** [comment]
The restore fix is now merged to master, with automatic tests that include backing up and restoring a long path.
If it is still possible to reproduce a problem after this fix, then we can reopen this issue.

Thank you for reporting, and everbody's suggestions.

**Comment by 4001982248998** [comment]
I have a new server with the aforementioned problematic behaviour: files with path names exceeding 259 characters are not saved, and throw “Unknown errors”. 

This is using burp 2.2.18 as client on Windows Server 2016 Version 1607 build 1439.3204, and on server it's burp 2.2.18 on Ubuntu Xenial 16.0.4.6 LTS.

I am now excluding the files by configuring “``exclude_regex = ^.{255,}$``”, but can change that easily on will when requested.

If I succeed learning how to record debugging logs with gdb or else, I will do so.

Do you want to reopen the issue, @grke ?

**Comment by grke** [comment]
Hello,

No, because the problem is fixed in burp-2.3.12.
